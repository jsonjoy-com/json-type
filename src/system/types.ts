import type {Schema, TypeOf} from '../schema';
import type {SchemaOf, Type} from '../type';
import type {TypeAlias} from './TypeAlias';

/** A custom user provided validator. */
export interface CustomValidator {
  /** Name of the validator. */
  name: string;

  /**
   * Receives the value that needs to be validated, returns a truthy value representing an error
   * or throws an error when validation fails. When validation succeeds, returns a falsy value.
   */
  fn: (value: any) => unknown;
}

export type TypeOfAlias<T> = T extends TypeAlias<any, infer T> ? T : T extends Type ? T : never;

export type ResolveType<T> =
  T extends TypeAlias<any, infer T>
    ? TypeOf<SchemaOf<T>>
    : T extends Type
      ? TypeOf<SchemaOf<T>>
      : T extends Schema
        ? TypeOf<T>
        : never;
