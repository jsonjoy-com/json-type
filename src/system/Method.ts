import type {TypeOf} from '../schema';
import type {SchemaOf} from '../type';
import type {TypeAlias, TypeOfAlias, TypeSystem} from '.';

export class Method<
  ID extends string,
  Req extends TypeAlias<any, any>,
  Res extends TypeAlias<any, any>,
  Ctx = unknown,
> {
  constructor(
    public readonly system: TypeSystem,
    public readonly id: ID,
    public readonly req: Req,
    public readonly res: Res,
    public readonly fn: (
      req: TypeOf<SchemaOf<TypeOfAlias<Req>>>,
      ctx: Ctx,
    ) => Promise<TypeOf<SchemaOf<TypeOfAlias<Res>>>>,
  ) {}
}
