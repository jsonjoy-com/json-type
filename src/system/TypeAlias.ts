import { printTree } from "tree-dump/lib/printTree";
import { ObjectType } from "../type/classes";
import { toText } from "../typescript/toText";
import type {
  JsonSchemaGenericKeywords,
  JsonSchemaValueNode,
} from "../json-schema";
import { typeToJsonSchema } from "../json-schema";
import { TypeExportContext } from "./TypeExportContext";
import type { TypeSystem } from ".";
import type { Type } from "../type";
import type { Printable } from "tree-dump/lib/types";
import type * as ts from "../typescript/types";

export class TypeAlias<K extends string, T extends Type> implements Printable {
  public constructor(
    public readonly system: TypeSystem,
    public readonly id: K,
    public readonly type: T,
  ) {}

  public getType(): Type {
    return this.type;
  }

  public resolve(): TypeAlias<string, Type> {
    return this.system.resolve(this.id);
  }

  public toString(tab: string = "") {
    return this.id + printTree(tab, [(tab) => this.type.toString(tab)]);
  }

  public toTypeScriptAst():
    | ts.TsInterfaceDeclaration
    | ts.TsTypeAliasDeclaration {
    const type = this.type;
    if (type instanceof ObjectType) {
      const ast = this.type.toTypeScriptAst() as ts.TsTypeLiteral;
      const node: ts.TsInterfaceDeclaration = {
        node: "InterfaceDeclaration",
        name: this.id,
        members: ast.members,
      };
      return node;
    } else {
      const node: ts.TsTypeAliasDeclaration = {
        node: "TypeAliasDeclaration",
        name: this.id,
        type: type.toTypeScriptAst(),
      };
      // TODO: Figure out if this is still needed, and possibly bring it back.
      // augmentWithComment(type, node);
      return node;
    }
  }

  public toTypeScript(): string {
    return toText(this.toTypeScriptAst());
  }

  public toJsonSchema(): JsonSchemaGenericKeywords {
    const node: JsonSchemaGenericKeywords = {
      $id: this.id,
      $ref: "#/$defs/" + this.id,
      $defs: {},
    };
    const ctx = new TypeExportContext();
    ctx.visitRef(this.id);
    node.$defs![this.id] = typeToJsonSchema(
      this.type,
      ctx,
    ) as JsonSchemaValueNode;
    let ref: string | undefined;
    while ((ref = ctx.nextMentionedRef())) {
      ctx.visitRef(ref);
      node.$defs![ref] = typeToJsonSchema(
        this.system.resolve(ref).type,
        ctx,
      ) as JsonSchemaValueNode;
    }
    return node;
  }
}
