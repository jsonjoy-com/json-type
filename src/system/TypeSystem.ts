import {RefType} from '../type/classes';
import {TypeBuilder} from '../type/TypeBuilder';
import {TypeAlias} from './TypeAlias';
import {printTree} from 'tree-dump/lib/printTree';
import {TypeMap} from '../schema';
import type {Type} from '../type';
import type {Printable} from 'tree-dump/lib/types';

/**
 * @todo Rename to Module.
 */
export class TypeSystem implements Printable {
  public readonly t = new TypeBuilder(this);

  public readonly aliases: Map<string, TypeAlias<string, any>> = new Map();

  /**
   * @todo Add ability fetch object of given type by its ID, analogous to
   * GraphQL "nodes".
   */
  public readonly alias = <K extends string, T extends Type>(id: K, type: T): TypeAlias<K, T> => {
    const existingAlias = this.aliases.get(id);
    if (existingAlias) return existingAlias as TypeAlias<K, T>;
    const alias = new TypeAlias<K, T>(this, id, type);
    this.aliases.set(id, alias);
    return alias;
  };

  public readonly unalias = <K extends string>(id: K): TypeAlias<K, Type> => {
    const alias = this.aliases.get(id);
    if (!alias) throw new Error(`Alias [id = ${id}] not found.`);
    return <TypeAlias<K, Type>>alias;
  };

  public readonly hasAlias = (id: string): boolean => this.aliases.has(id);

  public readonly resolve = <K extends string>(id: K): TypeAlias<K, Type> => {
    const alias = this.unalias(id);
    return alias.type instanceof RefType ? this.resolve<K>(alias.type.ref() as K) : alias;
  };

  public exportTypes() {
    const result: Record<string, unknown> = {};
    for (const [id, alias] of this.aliases.entries()) {
      result[id] = alias.getType().getSchema();
    }
    return result;
  }

  public importTypes<A extends TypeMap>(
    types: A,
  ): {
    readonly [K in keyof A]: TypeAlias<
      K extends string ? K : never,
      /** @todo Replace `any` by inferred type here. */ any
    >;
  } {
    const result = {} as any;
    for (const id in types) result[id] = this.alias(id, this.t.import(types[id]));
    return result;
  }

  public toString(tab: string = '') {
    return (
      'TypeSystem' +
      printTree(tab, [
        (tab) =>
          'aliases' +
          printTree(
            tab,
            [...this.aliases.values()].map((alias) => (tab) => alias.toString(tab)),
          ),
      ])
    );
  }
}
