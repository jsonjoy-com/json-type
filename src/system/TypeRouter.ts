import * as classes from '../type/classes';
import {TypeSystem} from './TypeSystem';
import {toText} from '../typescript/toText';
import type * as ts from '../typescript/types';
import type {TypeBuilder} from '../type/TypeBuilder';

export interface TypeRouterOptions<R extends RoutesBase> {
  system: TypeSystem;
  routes: R;
}

export class TypeRouter<Routes extends RoutesBase> {
  public static create = <NewRoutes extends RoutesBase>(
    routes?: (t: TypeRouter<{}>) => NewRoutes,
  ): TypeRouter<NewRoutes> => {
    const system = new TypeSystem();
    const router = new TypeRouter({
      system,
      routes: {},
    });
    return routes ? router.extend(routes) : (router as any);
  };

  public system: TypeSystem;
  public t: TypeBuilder;
  public routes: Routes;

  constructor(options: TypeRouterOptions<Routes>) {
    this.system = options.system;
    this.t = this.system.t;
    this.routes = options.routes;
    // this.system.importTypes(this.routes);
  }

  protected merge<Router extends TypeRouter<any>>(router: Router): TypeRouter<Routes & TypeRouterRoutes<Router>> {
    return new TypeRouter({
      system: this.system,
      routes: {
        ...this.routes,
        ...router.routes,
      },
    });
  }

  public extend<NewRoutes extends RoutesBase>(routes: (t: this) => NewRoutes): TypeRouter<Routes & NewRoutes> {
    const router = new TypeRouter({system: this.system, routes: routes(this)});
    return this.merge(router);
  }

  public fn<K extends string, R extends classes.FunctionType<any, any>>(
    name: K,
    type: R,
  ): TypeRouter<Routes & {[KK in K]: R}> {
    this.routes[name] = <any>type;
    return <any>this;
  }

  public fn$<K extends string, R extends classes.FunctionStreamingType<any, any>>(
    name: K,
    type: R,
  ): TypeRouter<Routes & {[KK in K]: R}> {
    this.routes[name] = <any>type;
    return <any>this;
  }

  public toTypeScriptAst(): ts.TsTypeLiteral {
    const node: ts.TsTypeLiteral = {
      node: 'TypeLiteral',
      members: [],
    };
    for (const [name, type] of Object.entries(this.routes)) {
      const schema = type.getSchema();
      const property: ts.TsPropertySignature = {
        node: 'PropertySignature',
        name,
        type: type.toTypeScriptAst(),
      };
      if (schema.title) property.comment = schema.title;
      node.members.push(property);
    }
    return node;
  }

  public toTypeScriptModuleAst(): ts.TsModuleDeclaration {
    const node: ts.TsModuleDeclaration = {
      node: 'ModuleDeclaration',
      name: 'Router',
      export: true,
      statements: [
        {
          node: 'TypeAliasDeclaration',
          name: 'Routes',
          type: this.toTypeScriptAst(),
          export: true,
        },
      ],
    };
    for (const alias of this.system.aliases.values()) node.statements.push({...alias.toTypeScriptAst(), export: true});
    return node;
  }

  public toTypeScript(): string {
    return toText(this.toTypeScriptModuleAst());
  }
}

export type RoutesBase = Record<string, classes.FunctionType<any, any> | classes.FunctionStreamingType<any, any>>;
type TypeRouterRoutes<R extends TypeRouter<any>> = R extends TypeRouter<infer R2> ? R2 : never;
