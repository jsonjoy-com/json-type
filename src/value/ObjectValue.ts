import {Value} from './Value';
import {toText} from '../typescript/toText';
import {TypeSystem} from '../system/TypeSystem';
import {printTree} from 'tree-dump/lib/printTree';
import type * as classes from '../type';
import type * as ts from '../typescript/types';
import type {TypeBuilder} from '../type/TypeBuilder';
import type {Printable} from 'tree-dump/lib/types';
import type {ResolveType} from '../system/types';

export type UnObjType<T> = T extends classes.ObjType<infer U> ? U : never;
export type UnObjectValue<T> = T extends ObjectValue<infer U> ? U : never;
export type UnObjectFieldTypeVal<T> = T extends classes.ObjectFieldType<any, infer U> ? U : never;
export type ObjectFieldToTuple<F> = F extends classes.ObjectFieldType<infer K, infer V> ? [K, V] : never;
export type ToObject<T> = T extends [string, unknown][] ? {[K in T[number] as K[0]]: K[1]} : never;
export type ObjectValueToTypeMap<F> = ToObject<{
  [K in keyof F]: ObjectFieldToTuple<F[K]>;
}>;
export type TuplesToFields<T> = T extends PropDefinition<infer K, infer V>[] ? classes.ObjectFieldType<K, V>[] : never;

type PropDefinition<K extends string, V extends classes.Type> = [key: K, val: V, data: ResolveType<V>];
type PropDef = <K extends string, V extends classes.Type>(key: K, val: V, data: ResolveType<V>) => PropDefinition<K, V>;

export class ObjectValue<T extends classes.ObjType<any>> extends Value<T> implements Printable {
  public static create = (system: TypeSystem = new TypeSystem()) => new ObjectValue(system.t.obj, {});

  public get system(): TypeSystem {
    return (this.type as T).getSystem();
  }

  public get t(): TypeBuilder {
    return this.system.t;
  }

  public keys(): string[] {
    const type = this.type as T;
    return type.fields.map((field: classes.ObjectFieldType<string, any>) => field.key);
  }

  public get<K extends keyof ObjectValueToTypeMap<UnObjType<T>>>(
    key: K,
  ): Value<
    ObjectValueToTypeMap<UnObjType<T>>[K] extends classes.Type ? ObjectValueToTypeMap<UnObjType<T>>[K] : classes.Type
  > {
    const field = this.type.getField(<string>key);
    if (!field) throw new Error('NO_FIELD');
    const type = field.value;
    const data = this.data[<string>key];
    return new Value(type, data) as any;
  }

  public field<F extends classes.ObjectFieldType<any, any>>(
    field: F | ((t: TypeBuilder) => F),
    data: ResolveType<UnObjectFieldTypeVal<F>>,
  ): ObjectValue<classes.ObjType<[...UnObjType<T>, F]>> {
    field = typeof field === 'function' ? field((this.type as classes.ObjType<any>).getSystem().t) : field;
    const extendedData = {...this.data, [field.key]: data};
    const type = this.type;
    const system = type.system;
    if (!system) throw new Error('NO_SYSTEM');
    const extendedType = system.t.Object(...type.fields, field);
    return new ObjectValue(extendedType, extendedData as any) as any;
  }

  public prop<K extends string, V extends classes.Type>(
    key: K,
    type: V | ((t: TypeBuilder) => V),
    data: ResolveType<V>,
  ) {
    const system = (this.type as classes.ObjType<any>).getSystem();
    const t = system.t;
    type = typeof type === 'function' ? type(t) : type;
    return this.field(t.prop(key, type), data);
  }

  public merge<O extends ObjectValue<any>>(
    obj: O,
  ): ObjectValue<classes.ObjType<[...UnObjType<T>, ...UnObjType<O['type']>]>> {
    const extendedData = {...this.data, ...obj.data};
    const type = this.type;
    const system = type.system;
    if (!system) throw new Error('NO_SYSTEM');
    const extendedType = system.t.Object(...type.fields, ...obj.type.fields);
    return new ObjectValue(extendedType, extendedData) as any;
  }

  public extend<R extends PropDefinition<any, any>[]>(
    inp: (t: TypeBuilder, prop: PropDef, system: TypeSystem) => R,
  ): ObjectValue<classes.ObjType<[...UnObjType<T>, ...TuplesToFields<R>]>> {
    const system = this.type.getSystem();
    const r: PropDef = (key, val, data) => [key, val, data];
    const extension = inp(system.t, r, system);
    const type = this.type;
    const extendedFields: classes.ObjectFieldType<any, any>[] = [...type.fields];
    const extendedData = {...this.data};
    for (const [key, val, data] of extension) {
      extendedFields.push(system.t.prop(key, val));
      extendedData[key] = data;
    }
    const extendedType = system.t.Object(...extendedFields);
    return new ObjectValue(extendedType, <any>extendedData) as any;
  }

  public toTypeScriptAst(): ts.TsTypeLiteral {
    const node: ts.TsTypeLiteral = {
      node: 'TypeLiteral',
      members: [],
    };
    const data = this.data as Record<string, classes.Type>;
    for (const [name, type] of Object.entries(data)) {
      const schema = type.getSchema();
      const property: ts.TsPropertySignature = {
        node: 'PropertySignature',
        name,
        type: type.toTypeScriptAst(),
      };
      if (schema.title) property.comment = schema.title;
      node.members.push(property);
    }
    return node;
  }

  public toTypeScriptModuleAst(): ts.TsModuleDeclaration {
    const node: ts.TsModuleDeclaration = {
      node: 'ModuleDeclaration',
      name: 'Router',
      export: true,
      statements: [
        {
          node: 'TypeAliasDeclaration',
          name: 'Routes',
          type: this.toTypeScriptAst(),
          export: true,
        },
      ],
    };
    const system = this.type.system;
    if (!system) throw new Error('system is undefined');
    for (const alias of system.aliases.values()) node.statements.push({...alias.toTypeScriptAst(), export: true});
    return node;
  }

  public toTypeScript(): string {
    return toText(this.toTypeScriptModuleAst());
  }

  public toString(tab: string = ''): string {
    return 'ObjectValue' + printTree(tab, [(tab) => this.type.toString(tab)]);
  }
}
