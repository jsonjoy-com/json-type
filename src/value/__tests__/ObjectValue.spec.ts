import {TypeSystem} from '../../json-type/system';
import {ObjectValue} from '../ObjectValue';

test('can retrieve field as Value', () => {
  const system = new TypeSystem();
  const {t} = system;
  const obj = new ObjectValue(t.Object(t.prop('foo', t.str)), {foo: 'bar'});
  const foo = obj.get('foo');
  expect(foo.type.getTypeName()).toBe('str');
  expect(foo.data).toBe('bar');
});

test('can print to string', () => {
  const system = new TypeSystem();
  const {t} = system;
  const obj = new ObjectValue(t.Object(t.prop('foo', t.str)), {foo: 'bar'});
  expect(obj + '').toMatchSnapshot();
});
