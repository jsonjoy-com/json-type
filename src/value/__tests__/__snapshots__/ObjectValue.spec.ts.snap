// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can print to string 1`] = `
"ObjectValue
└─ obj
   └─ "foo":
       └─ str"
`;
